experimental/dds/tree2/src/test/shared-tree/editing.spec.ts

```diff
@@ -1012,6 +1038,199 @@ describe("Editing", () => {
 
 				expectJsonTree([tree, tree2], [{ foo: "x" }]);
 			});
+
+			it("can put a constraint on newly inserted node", () => {
+				const tree = makeTreeFromJson([{}]);
+				const tree2 = tree.fork();
+
+				const rootPath: UpPath = {
+					parent: undefined,
+					parentField: rootFieldKeySymbol,
+					parentIndex: 0,
+				};
+
+				// Insert foo
+				// State should be: [{ foo: {}}]
+				const sequence = tree.editor.sequenceField({
+					parent: rootPath,
+					field: brand("foo"),
+				});
+				sequence.insert(0, singleTextCursor({ type: jsonObject.name }));
+
+				// Constrain on foo existinga and insert into root sequence if it does
+				// State should be: [{ foo: {}}, {}]
+				tree.transaction.start();
+				tree.editor.addNodeExistsConstraint({
+					parent: rootPath,
+					parentField: brand("foo"),
+					parentIndex: 0,
+				});
+				const rootSequence = tree.editor.sequenceField({
+					parent: undefined,
+					field: rootFieldKeySymbol,
+				});
+				rootSequence.insert(1, singleTextCursor({ type: jsonObject.name }));
+				tree.transaction.commit();
+
+				// Make a concurrent edit to rebase over that inserts into root sequence
+				// State should be (to tree2): [{}, {}]
+				const tree2RootSequence = tree2.editor.sequenceField({
+					parent: undefined,
+					field: rootFieldKeySymbol,
+				});
+				tree2RootSequence.insert(1, singleTextCursor({ type: jsonObject.name }));
+
+				tree.merge(tree2);
+				tree2.rebaseOnto(tree);
+
+				expectJsonTree([tree, tree2], [{ foo: {} }, {}, {}]);
+			});
+
+			it.skip("constraint on newly inserted node with ancestor deleted concurrently", () => {
+				const tree = makeTreeFromJson([{}]);
+				const tree2 = tree.fork();
+
+				const rootPath: UpPath = {
+					parent: undefined,
+					parentField: rootFieldKeySymbol,
+					parentIndex: 0,
+				};
+
+				// Insert "a"
+				// State should be: [{ foo: "a"}]
+				const sequence = tree.editor.sequenceField({
+					parent: rootPath,
+					field: brand("foo"),
+				});
+				sequence.insert(0, singleTextCursor({ type: jsonString.name, value: "a" }));
+
+				// Constrain on "a" existing and insert "b" into root sequence
+				// State should be (if "a" exists): [{ foo: "a"}, "b"]
+				// a's ancestor will be concurrently deleted
+				tree.transaction.start();
+				tree.editor.addNodeExistsConstraint({
+					parent: rootPath,
+					parentField: brand("foo"),
+					parentIndex: 0,
+				});
+				const rootSequence = tree.editor.sequenceField({
+					parent: undefined,
+					field: rootFieldKeySymbol,
+				});
+				rootSequence.insert(1, singleTextCursor({ type: jsonString.name, value: "b" }));
+				tree.transaction.commit();
+
+				// Concurrently delete a's ancestor
+				// State should be (to tree2): []
+				const tree2RootSequence = tree2.editor.sequenceField({
+					parent: undefined,
+					field: rootFieldKeySymbol,
+				});
+				tree2RootSequence.delete(0, 1);
```

Comment:
This edit has to be sequenced before the transaction with the constraint if we want the constraint to be violated. When we just have two blocks of concurrent edits I find it less confusing to use tree/tree1 as the branch whose edits come first and tree2 as the branch whose edits come second.

Reply:
But if you like you could keep the names and reverse the order of the merge/rebaseOnto calls below.

Reply:
Yeah I agree with your point about tree/tree1 having its edits first so I'll flip them. For some reason I thought this scenario should work with the order flipped and that we might need to make changes to support this.